### PR TITLE
Adding in metadata pydantic model and verification checks during runtime

### PR DESCRIPTION
**Pull request by snova-jayr**
Please refer to [CONTRIBUTING.md](CONTRIBUTING.md) for detailed instructions.

## PR Checklist
- [ ] My PR is less than 500 lines of code
- [ ] I have added sufficient comment as docstrings in my code
- [ ] I have made corresponding changes to the documentation

- [ ] I have written unit-tests to test all of my code

### CODE REVIEW

**Comment by snova-ranjanl** [comment]
@snova-jayr please add to PR description to justify the size of the changes in detail. Thank you!

File: generative_data_prep/utils/utils.py [hunk 1]
```diff
@@ -255,6 +260,92 @@ def execute_and_return_stdout(command):
     return result
 
 
+def _calculate_sha256(file_path: str):
+    sha256_hash = hashlib.sha256()
+    with open(file_path, "rb") as file:
+        for byte_block in iter(lambda: file.read(4096), b""):
+            sha256_hash.update(byte_block)
+    return sha256_hash.hexdigest()
```
Comment (snova-zoltanc) [inline generative_data_prep/utils/utils.py:268]: Crazy that this library does not have a builtin function to do this haha.

File: generative_data_prep/utils/utils.py [hunk 2]
```diff
@@ -255,6 +260,92 @@ def execute_and_return_stdout(command):
...
+    for foldername, _, filenames in os.walk(dir):
+        if filter is not None and filter in foldername.split("/"):
+            continue
+        relative_foldername = os.path.relpath(foldername, dir)
+        if relative_foldername == ".":
+            relative_foldername = ""
+        relative_foldername = relative_foldername.replace(os.path.sep, "_")
+        if relative_foldername != "":
+            relative_foldername += "_"
+        hash_file_name = [
```
Comment (snova-zoltanc) [inline generative_data_prep/utils/utils.py:282]: nit: keep lists variable names plural

File: README.md [hunk 3]
```diff
@@ -211,7 +211,71 @@ The metrics associated with this dataset will be printed in the terminal. These
...
+```
+max_seq_length: int
+token_type_ids: bool
+vocab_size: int
+tokenizer_model_type: str
+number_of_training_files: int
+number_of_dev_files: Optional[int]
+number_of_test_files: Optional[int]
+max_batch_size_train: int
+max_batch_size_dev: Optional[int]
```
Comment (snova-zoltanc) [inline README.md:227]: Curious question: Is evaluation done with batched inference?

File: generative_data_prep/utils/utils.py [hunk 4]
```diff
@@ -255,6 +260,92 @@ def execute_and_return_stdout(command):
...
+        current_modified_time = os.path.getmtime(file)
+        current_size = os.path.getsize(file)
+        if current_size != file_info_dict[hash_file_name]["size"] or not math.isclose(
+            current_modified_time, file_info_dict[hash_file_name]["modified_time"]
+        ):
+            file_hash = file_info_dict[hash_file_name]["sha256"]
+            current_file_hash = _calculate_sha256(file)
+            if file_hash != current_file_hash:
+                return False
+    return True
```
Comment (snova-zoltanc) [inline generative_data_prep/utils/utils.py:320]: nice! we will not call this during training though, right? My only worry is that if they want to customize the dataset by adding files or something it would fail out?

File: generative_data_prep/utils/metadata_generation.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        return v
+
+    @field_validator("vocab_size")
+    @classmethod
+    def validation_vocab_size(cls, v: int, info: ValidationInfo):
+        """Validates vocab size."""
+        runtime_vocab_size = info.context.get("vocab_size")
+        if type(runtime_vocab_size) is not int:
+            raise ValueError("Vocab size context param should be an integer variable")
+        if v > runtime_vocab_size:
```
Comment (snova-zoltanc) [inline generative_data_prep/utils/metadata_generation.py:94]: greater than or equal to?

File: generative_data_prep/utils/metadata_generation.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        return v
+
+    @field_validator("number_of_training_files")
+    @classmethod
+    def validate_number_of_training_files(cls, v: int, info: ValidationInfo):
+        """Validates number of training files."""
+        number_of_instances = info.context.get("world_size")
+        if type(number_of_instances) is not int:
+            raise ValueError("World size context param should be an integer variable")
+        if v < number_of_instances:
```
Comment (snova-zoltanc) [inline generative_data_prep/utils/metadata_generation.py:78]: The number of workers can be greater than the number of files?

v >= number_of_instances:

Like you can run data parallel 16 on 64 workers.

File: generative_data_prep/utils/metadata_generation.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                    f"must match max_seq_length used during generative data prep ({v})"
+                )
+            )
+        return v
+
+    @field_validator("number_of_training_files")
+    @classmethod
+    def validate_number_of_training_files(cls, v: int, info: ValidationInfo):
+        """Validates number of training files."""
+        number_of_instances = info.context.get("world_size")
```
Comment (snova-zoltanc) [inline generative_data_prep/utils/metadata_generation.py:75]: nit: why change variable name?

**Review by snova-zoltanc — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -211,7 +211,71 @@ The metrics associated with this dataset will be printed in the terminal. These
...
+To help improve speed and cross-checking we now provide a metadata file along with the dataset. This file is located right under the `output_dir` as `metadata.yaml`. This file is used along with a custom pydantic model which you can import from this library which will verify the dataset parameters and the training parameters. This can be used as a way to catch bugs before training begins.
+#### Structure of Metadata file
+```
+max_seq_length: int
+token_type_ids: bool
+vocab_size: int
+tokenizer_model_type: str
+number_of_training_files: int
+number_of_dev_files: Optional[int]
+number_of_test_files: Optional[int]
```
[the remark this answers is not in the document]
Comment (snova-jayr) [reply README.md:225]: should not be necessary! But I guess left that in there for now in case need to use it for something in the future or if you want to see if there are test data files in the dataset!

File: README.md [hunk 9]
```diff
@@ -211,7 +211,71 @@ The metrics associated with this dataset will be printed in the terminal. These
...
+max_batch_size_train: int
+max_batch_size_dev: Optional[int]
+```
+
+NOTE:
+* `tokenizer_model_type` is the string conversion of `type(modelConfig)`. Can use this field to compare the model used during training, which can be extracted by using `AutoConfig` in Huggingface transformers. Then wrapping it with `str(type())`.
+* `number_of_dev_files` and `max_batch_size_dev` will be `None` unless dev files are created during generative data pipeline.
+* `number_of_test_files` will be `None` unless test files are created during generative data pipeline.
+* `token_type_ids` will always be `True` for now since they are always generated.
```
[the remark this answers is not in the document]
Comment (snova-jayr) [reply README.md:234]: added it to verify in pydantic model! And I guess if it needs to be changed in the future it can be there!

File: README.md (hunk 3 above)
Comment (snova-jayr) [reply README.md:227]: This I will check on in the gpt2 task file

File: generative_data_prep/utils/utils.py (hunk 4 above)
Comment (snova-jayr) [reply generative_data_prep/utils/utils.py:320]: hmmmm I guess we could have some way where if a user wants to add some new files to the dataset we can have a function that can take in the new path and generate the sha256 and add it to the hash file? Does that sound like a good idea to you? And I guess this is more of a failsafe, so if it passes the inital metadata then it will not scrub the dataset by looking into each hdf5 file, but if it fails the metadata check then it will do the current runtime checks we do where it looks into each file!

File: generative_data_prep/utils/metadata_generation.py (hunk 5 above)
Comment (snova-jayr) [reply generative_data_prep/utils/metadata_generation.py:94]: or it should be ok to have runtime vocab size = dataset vocab size correct?

File: generative_data_prep/utils/metadata_generation.py (hunk 6 above)
Comment (snova-jayr) [reply generative_data_prep/utils/metadata_generation.py:78]: or can the number of workers be greater than the number of files? I thought that there has to be at least 64 files in order for each worker to get a file or am I missing something here

File: generative_data_prep/utils/metadata_generation.py (hunk 7 above)
Comment (snova-jayr) [reply generative_data_prep/utils/metadata_generation.py:75]: hmmm not too sure what the runtime variable would be called LOL at least world size in our context makes sense, maybe instead of world_size I can call it number_of_workers so its more generic!

File: generative_data_prep/data_prep/pipeline.py [hunk 10]
```diff
@@ -290,10 +298,23 @@ def multiprocess_data_prep(
     broken_process_indices = []
     broken_process_pool_exc: Optional[BaseException] = None
     metrics = Metrics()
+    max_batch_size_train = None
+    max_batch_size_dev = None
     # search for any "interesting" exception (a non-BrokenProcessPool Exception)
     for i, future in enumerate(futures):
         try:
-            metrics += future.result()
+            indiv_metric = future.result()
```
[the remark this answers is not in the document]
Comment (snova-jayr) [reply generative_data_prep/data_prep/pipeline.py:306]: ahhh I guess individual to indicate that this is the metric that pertains to a single file!

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (snova-zoltanc) [reply README.md:227]: Ok makes sense, I general like to not add unused code until it is necessary, but this sounds good

File: generative_data_prep/utils/metadata_generation.py (hunk 5 above)
Comment (snova-zoltanc) [reply generative_data_prep/utils/metadata_generation.py:101]: I am dumb! this looks good

File: generative_data_prep/utils/metadata_generation.py (hunk 6 above)
Comment (snova-zoltanc) [reply generative_data_prep/utils/metadata_generation.py]: This looks good I was being dumb. sorry

File: generative_data_prep/utils/utils.py (hunk 4 above)
Comment (snova-zoltanc) [reply generative_data_prep/utils/utils.py:320]: Not sure I quite understand, but I trust.

Another option is to just fail out during training and allow the user to over-ride the checking with a flag like --do_not_verify_input_data if they customize it?

Comment (snova-jayr) [reply generative_data_prep/utils/utils.py:320]: ah I guess its kind of like this

use sha256 to verify that none of the files have been modified:

if the files have been modified then we do stage 1 (going through each hdf5 files and extracting the necessary data components which should take a while) https://github.sambanovasystems.com/SambaNova/software/blob/151953ccb83dadb5ad70986d2f054a404dd87584/sambaflow/apps/nlp/transformers_on_rdu/transformers_hook_v2.py#L168-L171

if the files have not been modified then we do stage 2 (this PR where we just check using the metadata file)

but you're right in that we should have some type of override check, I guess for sambastudio what I imagined is that they could do stage 2 in SS suite before even alloc resources so you dont have to wait for RDUs to open up, and if it passes then we can pass a flag to the modelbox saying there is no need to do stage 1!

File: generative_data_prep/utils/metadata_generation.py [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+            if runtime_batch_size > v:
+                raise ValueError(
+                    (
+                        f"batch size specified during training ({runtime_batch_size}) exceeds the maximum "
+                        f"allowed batch size ({v}) based on evaluation files in dataset"
+                    )
+                )
+
+
+if __name__ == "__main__":
```
[the remark this answers is not in the document]
Comment (snova-jayr) [reply generative_data_prep/utils/metadata_generation.py]: deleted

File: generative_data_prep/utils/utils.py (hunk 4 above)
Comment (snova-zoltanc) [reply generative_data_prep/utils/utils.py:320]: This makes perfect sense now! 99% of the time we should be able to work with stage 2! 

Make sure that you account for the fact that users can copy in new files, not just modify the existing files

**Review by snova-zoltanc — CHANGES_REQUESTED** [review]
Remeber to update pyproject.toml with pydantic

**Review by snova-ranjanl — CHANGES_REQUESTED** [review]
Blocking it just for the size of files; let's review this together to talk through and identify if it's needed or if we can figure out ways to reduce it.

File: tests/component/test_running_command.py [hunk 12]
```diff
@@ -18,7 +18,7 @@ def test_pipeline_runs_successfully(run_path):
     THEN the pipeline runs successfully
         And returns a return code of 0 within specified timeout
     """
-    time_limit_seconds = 9
```
Comment (snova-ranjanl) [inline tests/component/test_running_command.py:21]: We need a `TODO` here to reduce this time in future PR

Comment (snova-jayr) [reply tests/component/test_running_command.py:21]: done!

**Review by snova-ranjanl — APPROVED** [review]

File: README.md [hunk 13]
```diff
@@ -211,6 +211,75 @@ The metrics associated with this dataset will be printed in the terminal. These
...
+Here is an example code of what this would look like!
+
+```
+from generative_data_prep.utils import validate_sha256
+
+validate_sha256(output_dir)
+```
+`output_dir` here should point to the directory which was created using generative data pipeline. This function returns a `bool` and will be `True` if there is NO corruption in the dataset and `False` if there is corruption in the dataset.
+
+Under the hood each file is scrubbed through and is first checked with the size and modified date. If these values are not the same as when the file was first created then the function will calculate the sha256 of the file and compare it to what is saved.
```
Comment (snova-zoltanc) [inline README.md:280]: Just to confirm, there is no way that the file can change without updating the modified date? So there is no way the sha will change if the modified date did not change?

File: generative_data_prep/data_prep/pipeline.py [hunk 14]
```diff
@@ -384,6 +409,12 @@ def pipeline_main(  # noqa: C901
         Metrics associated with tokenization
     """
     # print input file information
+    dataset_metadata_json = {
```
Comment (snova-zoltanc) [inline generative_data_prep/data_prep/pipeline.py:412]: nit: this should be an object, either using the "DatasetMetadata"  or in corporate it into the existing metrics class? It is bad practice to create objects in python using dictionaries. Especially when the attributes are so well defined like here (you know the types and such)  

We can do this in a future PR, if you agree then just open an issue and then create a PR for it in the future.

File: tests/test_metadata.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+        "--pretrained_tokenizer",
+        "gpt2",
+        "--max_seq_length",
+        "1024",
+        "--shuffle",
+        "on_RAM",
+        "--num_training_splits",
+        "4",
+    ]
+    subprocess.run(command + ["--keep_split_jsonls", "--output_path", f"{tmp_e2e_output_dir}"])
```
Comment (snova-zoltanc) [inline tests/test_metadata.py:276]: This might be risky depending on if user has python3 defined, please update it to call pipeline_main directly

as in file below

https://github.com/sambanova/generative_data_prep/blob/main/tests/test_e2e.py


image

File: tests/test_metadata.py [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+        "model_type": str(type(AutoConfig.from_pretrained("gpt2"))),
+        "use_token_type_ids": use_token_type_ids,
+        "vocab_size": 50257,
+        "number_of_workers": 4,
+        "max_seq_length": 1024,
+    }
+    DatasetMetadata.model_validate(metadata_dict, context=context_dict)
+
+
+def test_metadata_end2end_output(tmp_path):
```
Comment (snova-zoltanc) [inline tests/test_metadata.py:253]: Awesome test, thanks!

File: tests/test_sha256.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        "--pretrained_tokenizer",
+        "gpt2",
+        "--max_seq_length",
+        "1024",
+        "--shuffle",
+        "on_RAM",
+        "--num_training_splits",
+        "4",
+    ]
+    subprocess.run(
```
Comment (snova-zoltanc) [inline tests/test_sha256.py:87]: same as above, call pipeline main directly if possible

File: tests/test_sha256.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        "--pretrained_tokenizer",
+        "gpt2",
+        "--max_seq_length",
+        "1024",
+        "--shuffle",
+        "on_RAM",
+        "--num_training_splits",
+        "4",
+    ]
+    subprocess.run(command + ["--keep_split_jsonls", "--output_path", f"{tmp_split_output_dir}"])
```
Comment (snova-zoltanc) [inline tests/test_sha256.py:36]: same as above, call pipeline main

File: tests/test_sha256.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        "--pretrained_tokenizer",
+        "gpt2",
+        "--max_seq_length",
+        "1024",
+        "--shuffle",
+        "on_RAM",
+        "--num_training_splits",
+        "4",
+    ]
+    subprocess.run(command + ["--output_path", f"{tmp_no_split_output_dir}"])
```
Comment (snova-zoltanc) [inline tests/test_sha256.py:126]: same as above, call pipeline main directly

**Review by snova-zoltanc — DISMISSED** [review]
Please update the test cases to not use subprocess, and import and call main functions directly
